File: app/src/main/java/org/wikipedia/feed/view/CardHeaderView.java
package org.wikipedia.feed.view;

import android.content.Context;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.view.Gravity;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.appcompat.widget.PopupMenu;
import androidx.constraintlayout.widget.ConstraintLayout;

import org.wikipedia.R;
import org.wikipedia.WikipediaApp;
import org.wikipedia.feed.model.Card;
import org.wikipedia.util.DimenUtil;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class CardHeaderView extends ConstraintLayout {
    public interface Callback {
        boolean onRequestDismissCard(@NonNull Card card);
        void onRequestEditCardLanguages(@NonNull Card card);
        void onRequestCustomize(@NonNull Card card);
    }

    @BindView(R.id.view_card_header_title) TextView titleView;
    @BindView(R.id.view_list_card_header_lang_background) View langCodeBackground;
    @BindView(R.id.view_list_card_header_lang_code) TextView langCodeView;
    @Nullable private Card card;
    @Nullable private Callback callback;
    private static final int PADDING = 4;

    public CardHeaderView(Context context) {
        super(context);
        init();
    }

    public CardHeaderView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public CardHeaderView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        inflate(getContext(), R.layout.view_card_header, this);
        ButterKnife.bind(this);
        setPadding(0, 0, 0, DimenUtil.roundedDpToPx(PADDING));
    }

    @NonNull public CardHeaderView setCard(@NonNull Card card) {
        this.card = card;
        return this;
    }

    @NonNull public CardHeaderView setCallback(@Nullable Callback callback) {
        this.callback = callback;
        return this;
    }

    @NonNull public CardHeaderView setTitle(@Nullable CharSequence title) {
        titleView.setText(title);
        return this;
    }

    @NonNull public CardHeaderView setTitle(@StringRes int id) {
        titleView.setText(id);
        return this;
    }

    @NonNull public CardHeaderView setLangCode(@Nullable String langCode) {
        if (TextUtils.isEmpty(langCode) || WikipediaApp.getInstance().language().getAppLanguageCodes().size() < 2) {
            langCodeBackground.setVisibility(GONE);
            langCodeView.setVisibility(GONE);
        } else {
            langCodeBackground.setVisibility(VISIBLE);
            langCodeView.setVisibility(VISIBLE);
            langCodeView.setText(langCode);
        }
        return this;
    }

    @OnClick(R.id.view_list_card_header_menu) void onMenuClick(View v) {
        showOverflowMenu(v);
    }

    private void showOverflowMenu(View anchorView) {
        PopupMenu menu = new PopupMenu(anchorView.getContext(), anchorView, Gravity.END);
        menu.getMenuInflater().inflate(R.menu.menu_feed_card_header, menu.getMenu());
        MenuItem editCardLangItem = menu.getMenu().findItem(R.id.menu_feed_card_edit_card_languages);
        editCardLangItem.setVisible(card.type().contentType().isPerLanguage());
        menu.setOnMenuItemClickListener(new CardHeaderMenuClickListener());
        menu.show();
    }

    private class CardHeaderMenuClickListener implements PopupMenu.OnMenuItemClickListener {
        @Override
        public boolean onMenuItemClick(MenuItem item) {
            switch (item.getItemId()) {
                case R.id.menu_feed_card_dismiss:
                    if (callback != null & card != null) {
                        return callback.onRequestDismissCard(card);
                    }
                    return false;

                case R.id.menu_feed_card_edit_card_languages:
                    if (callback != null & card != null) {
                        callback.onRequestEditCardLanguages(card);
                    }
                    return true;

                case R.id.menu_feed_card_customize:
                    if (callback != null & card != null) {
                        callback.onRequestCustomize(card);
                    }
                    return true;
                default:
                    return false;
            }
        }
    }
}

File: app/src/main/java/org/wikipedia/feed/view/CardHeaderView.kt
package org.wikipedia.feed.view

import android.content.Context
import android.util.AttributeSet
import android.view.Gravity
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import androidx.annotation.StringRes
import androidx.appcompat.widget.PopupMenu
import androidx.constraintlayout.widget.ConstraintLayout
import org.wikipedia.R
import org.wikipedia.WikipediaApp
import org.wikipedia.databinding.ViewCardHeaderBinding
import org.wikipedia.feed.model.Card
import org.wikipedia.util.L10nUtil

class CardHeaderView constructor(context: Context, attrs: AttributeSet? = null) : ConstraintLayout(context, attrs) {

    interface Callback {
        fun onRequestDismissCard(card: Card): Boolean
        fun onRequestEditCardLanguages(card: Card)
        fun onRequestCustomize(card: Card)
    }

    private val binding = ViewCardHeaderBinding.inflate(LayoutInflater.from(context), this)

    private var card: Card? = null
    private var callback: Callback? = null
    var titleView = binding.viewCardHeaderTitle
        private set

    init {
        binding.viewListCardHeaderMenu.setOnClickListener { showOverflowMenu(it) }
    }

    private fun showOverflowMenu(anchorView: View) {
        card?.let {
            val menu = PopupMenu(anchorView.context, anchorView, Gravity.END)
            menu.menuInflater.inflate(R.menu.menu_feed_card_header, menu.menu)
            val editCardLangItem = menu.menu.findItem(R.id.menu_feed_card_edit_card_languages)
            editCardLangItem.isVisible = it.type().contentType()?.run { isPerLanguage } ?: false
            menu.setOnMenuItemClickListener(CardHeaderMenuClickListener())
            menu.show()
        }
    }

    fun setCard(card: Card): CardHeaderView {
        this.card = card
        return this
    }

    fun setCallback(callback: Callback?): CardHeaderView {
        this.callback = callback
        return this
    }

    fun setTitle(title: CharSequence?): CardHeaderView {
        binding.viewCardHeaderTitle.text = title
        return this
    }

    fun setTitle(@StringRes id: Int): CardHeaderView {
        binding.viewCardHeaderTitle.setText(id)
        return this
    }

    fun setLangCode(langCode: String?): CardHeaderView {
        if (langCode.isNullOrEmpty() || WikipediaApp.instance.appLanguageState.appLanguageCodes.size < 2) {
            binding.viewListCardHeaderLangBackground.visibility = View.GONE
            binding.viewListCardHeaderLangCode.visibility = View.GONE
            L10nUtil.setConditionalLayoutDirection(this, WikipediaApp.instance.appLanguageState.systemLanguageCode)
        } else {
            binding.viewListCardHeaderLangBackground.visibility = VISIBLE
            binding.viewListCardHeaderLangCode.visibility = VISIBLE
            binding.viewListCardHeaderLangCode.text = langCode
            L10nUtil.setConditionalLayoutDirection(this, langCode)
        }
        return this
    }

    private inner class CardHeaderMenuClickListener : PopupMenu.OnMenuItemClickListener {
        override fun onMenuItemClick(item: MenuItem): Boolean {
            return card?.let {
                when (item.itemId) {
                    R.id.menu_feed_card_dismiss -> {
                        callback?.onRequestDismissCard(it)
                        true
                    }
                    R.id.menu_feed_card_edit_card_languages -> {
                        callback?.onRequestEditCardLanguages(it)
                        true
                    }
                    R.id.menu_feed_card_customize -> {
                        callback?.onRequestCustomize(it)
                        true
                    }
                    else -> false
                }
            } ?: run { false }
        }
    }
}
